routers/taskRoutes.js

```diff
@@ -0,0 +1,221 @@
+ * Allowed filters { title, completed, priority, deadlineDate}
+ *
+ * Throws UnauthorizedError or NotFoundError (based on user privileges) if no userID found
+ *
+ * Authorization required: logged in as correct user or admin */
+
+//PEER Lawrence, below I process authentication (ensureLoggedIn) before schema validation (validateRequest) and authorization (userExistsAndCorrect).
+//Would you prefer I change this order? I personally think that running ensureLoggedIn first is okay because it's a lighter/faster operation than validateRequest and userExistsAndCorrect,
+// and it makes sense to me to ensure the user is logged in before schema validation.
+router.get(
+  "/:userId",
+  ensureLoggedIn,
+  validateRequest([
+    { schema: taskGetSchema, userIdParam: true, reqQuery: true },
+  ]),
+  userExistsAndCorrect,
+  async function (req, res, next) {
+    try {
+      // Handle request with filters
+      if (Object.keys(req.query).length > 0) {
+        const tasks = await Task.getByFilters({
+          userId: req.params.userId,
+          ...req.query,
+        });
+        return res.status(200).json({ tasks });
+      }
+      const tasks = await Task.getAll(req.params.userId);
+      return res.status(200).json({ tasks });
+    } catch (error) {
+      // console.log("caught error");
+      // console.log(error);
+      return next(error);
+    }
+  }
+);
+
+/** POST /userId { taskData } =>  { task }
+ *
+ * Creates new task.
+ *
+ * taskData required fields: { title, urgent, important }
+ *
+ * taskData optional fields for 'taskData': { timebox, note, category, deadlineDate}
+ *
+ * Returns {task: {taskId, userId, title, urgent, important, priority, timebox, completed, note, category, deadlineDate }}
+ *
+ * Throws UnauthorizedError or NotFoundError (based on user privileges) if no userID found
+ *
+ * Authorization required: logged in as correct user or admin */
+
+router.post(
+  "/:userId",
+  ensureLoggedIn,
+  validateRequest([
+    { schema: taskCreateSchema, userIdParam: true, reqBody: true },
+  ]),
+  userExistsAndCorrect,
+  async function (req, res, next) {
+    try {
+      const task = await Task.create({
+        userId: req.params.userId,
+        ...req.body,
+      });
+      return res.status(201).json({ task });
+    } catch (error) {
+      // console.log("error from POST /task/:userId");
+      // console.log(error);
+      return next(error);
+    }
+  }
+);
+
+/** PATCH /:userId/:taskId { newData } =>  { task }
+ *
+ * Updates a task based on fields provided.
+ *
+ * newData optional fields: { title, urgent, important, timebox, completed, note, category, deadline_date }.
+ *
+ * Returns {task: {taskId, userId, fieldChanged1, fieldChanged2 ... }}
+ *
+ * Throws UnauthorizedError or NotFoundError (based on user privileges) if no taskID or userID found
+ *
+ * Authorization required: logged in as correct user or admin */
+
+/** //PEER Lawrence, which route hierarchy is better:
+current:
+-  patch(tasks/:userId/:taskId) 
```

Comment:
This depends on the entity relationship, but will talk more about it on our call